src/mlpack/methods/ann/layer/grouped_convolution_impl.hpp

```diff
@@ -405,60 +432,42 @@ void GroupedConvolutionType<
     const size_t fullInputOffset = offset * inMaps;
     const size_t fullOutputOffset = offset * maps;
 
+    #pragma omp parallel for collapse(2)
     for (size_t group = 0; group < groups; group++)
     {
-    // Iterate over input maps.
+      // Iterate over input maps.
       for (size_t inMap = 0; inMap < inGroupSize; ++inMap)
       {
+        MatType output;
         // Iterate over output maps.
         for (size_t outMap = group * outGroupSize; 
             outMap < ((group + 1) * outGroupSize); ++outMap)
         {
-          MatType output;
-
           BackwardConvolutionRule::Convolution(
-              mappedError.slice(outMap + fullOutputOffset),
+              dilatedMappedError.slice(outMap + fullOutputOffset),
               rotatedFilters.slice((outMap * inGroupSize) + inMap),
               output,
-              strideHeight,
-              strideWidth);
-
-          // If the stride width or height is greater than 1, then we have to
-          // insert columns and rows into the convolution output.
-          if (strideWidth == 1 && strideHeight == 1)
-          {
-            if (usingPadding)
-            {
-              gTemp.slice((group * inGroupSize) + inMap + fullInputOffset) +=
-                  output.submat(
-                      padWLeft,
-                      padHTop,
-                      padWLeft + gTemp.n_rows - 1,
-                      padHTop + gTemp.n_cols - 1);
-            }
-            else
-            {
-              gTemp.slice((group * inGroupSize) + inMap + fullInputOffset) +=
-                  output;
-            }
-          }
-          else
-          {
-            // We must iterate over each element of the output and manually
-            // re-insert the stride.
-            size_t col = padWLeft;
-            for (size_t i = 0; i < output.n_cols; ++i)
-            {
-              size_t row = padHTop;
-              for (size_t j = 0; j < output.n_rows; ++j)
-              {
-                gTemp(row, col, (group * inGroupSize) + inMap +
-                    fullInputOffset) += output(j, i);
-                row += strideHeight;
-              }
-              col += strideWidth;
-            }
-          }
+              1,
+              1,
+              1,
+              1,
+              outMap > group * outGroupSize);
+        }
+        // If the stride width or height is greater than 1, then we have to
+        // insert columns and rows into the convolution output.
+        MatType& curGTemp = gTemp.slice((group * inGroupSize) + inMap +
+            fullInputOffset);
+        if (usingPadding)
+        {
+          curGTemp = output.submat(
+              padWLeft,
+              padHTop,
+              padWLeft + gTemp.n_rows - 1,
+              padHTop + gTemp.n_cols - 1);
+        }
+        else
+        {
+          curGTemp = output;
```

Comment:
It's possible that an extra optimization could be to use `gTemp` directly when `usingPadding` is false, but we can leave that for later.